File: src/sec/ticker-map.ts
import { CLIError, ErrorCode } from '../core/errors.js';
import { tickerMapUrl } from './endpoints.js';
import { isLikelyCik, normalizeCik, normalizeTicker } from './normalizers.js';
import { SecClient } from './client.js';

export interface TickerRecord {
  cik_str: number;
  ticker: string;
  title: string;
}

export interface ResolvedEntity {
  input: string;
  cik: string;
  cik_numeric: number;
  ticker: string | null;
  title: string | null;
}

let cachedMap: TickerRecord[] | null = null;
let cachedAtMs = 0;
const CACHE_TTL_MS = 15 * 60 * 1000;

async function getTickerMap(client: SecClient): Promise<TickerRecord[]> {
  const now = Date.now();
  if (cachedMap && now - cachedAtMs < CACHE_TTL_MS) {
    return cachedMap;
  }

  const payload = await client.fetchSecJson<Record<string, TickerRecord>>(tickerMapUrl());
  const records = Object.values(payload)
    .filter((record) => record && typeof record === 'object')
    .filter((record) => typeof record.cik_str === 'number' && typeof record.ticker === 'string');

  cachedMap = records;
  cachedAtMs = now;

  return records;
}

function findByTicker(records: TickerRecord[], ticker: string): TickerRecord | undefined {
  return records.find((record) => record.ticker.toUpperCase() === ticker);
}

function findByCik(records: TickerRecord[], cik10: string): TickerRecord | undefined {
  const cikNumeric = Number.parseInt(cik10, 10);
  return records.find((record) => record.cik_str === cikNumeric);
}

export async function resolveEntity(
  id: string,
  client: SecClient,
  options?: { strictMapMatch?: boolean }
): Promise<ResolvedEntity> {
  const strictMapMatch = options?.strictMapMatch ?? false;
  const records = await getTickerMap(client);

  if (isLikelyCik(id)) {
    const cik = normalizeCik(id);
    const cikNumeric = Number.parseInt(cik, 10);
    const match = findByCik(records, cik);

    if (!match && strictMapMatch) {
      throw new CLIError(ErrorCode.NOT_FOUND, `No SEC ticker-map record found for CIK ${cik}`);
    }

    return {
      input: id,
      cik,
      cik_numeric: cikNumeric,
      ticker: match?.ticker ?? null,
      title: match?.title ?? null
    };
  }

  const ticker = normalizeTicker(id);
  const match = findByTicker(records, ticker);

  if (!match) {
    throw new CLIError(ErrorCode.NOT_FOUND, `No SEC ticker-map record found for ticker ${ticker}`);
  }

  return {
    input: id,
    cik: String(match.cik_str).padStart(10, '0'),
    cik_numeric: match.cik_str,
    ticker: match.ticker,
    title: match.title
  };
}

File: pkg/edgar/ticker_map.go
package edgar

import (
	"context"
	"strconv"
	"strings"
	"sync"
	"time"
)

type TickerRecord struct {
	CIKStr int    `json:"cik_str"`
	Ticker string `json:"ticker"`
	Title  string `json:"title"`
}

type ResolvedEntity struct {
	Input      string  `json:"input"`
	CIK        string  `json:"cik"`
	CIKNumeric int     `json:"cik_numeric"`
	Ticker     *string `json:"ticker"`
	Title      *string `json:"title"`
}

type tickerMapCache struct {
	mu        sync.Mutex
	records   []TickerRecord
	cachedAt  time.Time
	cacheHost string
}

var globalTickerCache tickerMapCache

const tickerMapTTL = 15 * time.Minute

func getTickerMap(ctx context.Context, client *SecClient) ([]TickerRecord, error) {
	globalTickerCache.mu.Lock()
	if len(globalTickerCache.records) > 0 &&
		time.Since(globalTickerCache.cachedAt) < tickerMapTTL &&
		globalTickerCache.cacheHost == client.Hosts().WWW {
		records := append([]TickerRecord(nil), globalTickerCache.records...)
		globalTickerCache.mu.Unlock()
		return records, nil
	}
	globalTickerCache.mu.Unlock()

	var payload map[string]TickerRecord
	if err := client.FetchSECJSON(ctx, tickerMapURL(client.Hosts()), &payload); err != nil {
		return nil, err
	}
	records := make([]TickerRecord, 0, len(payload))
	for _, record := range payload {
		if record.CIKStr != 0 && record.Ticker != "" {
			records = append(records, record)
		}
	}

	globalTickerCache.mu.Lock()
	globalTickerCache.records = append([]TickerRecord(nil), records...)
	globalTickerCache.cachedAt = time.Now()
	globalTickerCache.cacheHost = client.Hosts().WWW
	globalTickerCache.mu.Unlock()
	return records, nil
}

func resetTickerMapCache() {
	globalTickerCache.mu.Lock()
	defer globalTickerCache.mu.Unlock()
	globalTickerCache.records = nil
	globalTickerCache.cachedAt = time.Time{}
	globalTickerCache.cacheHost = ""
}

func resolveEntity(ctx context.Context, id string, client *SecClient, strictMapMatch bool) (ResolvedEntity, error) {
	records, err := getTickerMap(ctx, client)
	if err != nil {
		return ResolvedEntity{}, err
	}

	if isLikelyCIK(id) {
		cik, err := normalizeCIK(id)
		if err != nil {
			return ResolvedEntity{}, err
		}
		cikNumeric, _ := strconv.Atoi(cik)
		var match *TickerRecord
		for idx := range records {
			if records[idx].CIKStr == cikNumeric {
				match = &records[idx]
				break
			}
		}
		if match == nil && strictMapMatch {
			return ResolvedEntity{}, NewCLIError(ErrorNotFound, "No SEC ticker-map record found for CIK "+cik)
		}
		var ticker *string
		var title *string
		if match != nil {
			ticker = &match.Ticker
			title = &match.Title
		}
		return ResolvedEntity{
			Input:      id,
			CIK:        cik,
			CIKNumeric: cikNumeric,
			Ticker:     ticker,
			Title:      title,
		}, nil
	}

	ticker, err := normalizeTicker(id)
	if err != nil {
		return ResolvedEntity{}, err
	}
	for _, record := range records {
		if strings.ToUpper(record.Ticker) == ticker {
			cik := strconv.Itoa(record.CIKStr)
			return ResolvedEntity{
				Input:      id,
				CIK:        strings.Repeat("0", 10-len(cik)) + cik,
				CIKNumeric: record.CIKStr,
				Ticker:     &record.Ticker,
				Title:      &record.Title,
			}, nil
		}
	}

	return ResolvedEntity{}, NewCLIError(ErrorNotFound, "No SEC ticker-map record found for ticker "+ticker)
}
